File: core/router.py
"""Router: turns free text or form input into a structured TripRequest.

Two paths:
1. `route_from_form(...)` — direct mapping from the sidebar form (no LLM).
2. `route_from_text(...)` — LLM structured-output extraction from chat text
   (pydantic schema via LangChain `with_structured_output`), with a
   rule-based regex fallback when the LLM is unavailable.
"""
from __future__ import annotations

import re
from datetime import date, timedelta
from typing import Any

from pydantic import BaseModel, Field

from core.state import TripRequest

_SYSTEM_PROMPT = (
    "You are a travel planning assistant. Extract trip details from the user's "
    "message into the TripRequest schema. Use today's date "
    f"({date.today().isoformat()}) as reference when interpreting relative dates "
    "like 'next week'. If a date is ambiguous, pick the nearest sensible future "
    "date. Budget should be a number in USD. Travelers defaults to 1."
)


class _TripRequestExtraction(BaseModel):
    """Minimal schema for LLM extraction (no display-only fields).

    Keeping the schema small avoids confusing small local models, which tend
    to stuff values into extra fields (e.g. putting 'London' into a display
    field instead of `origin`).
    """

    origin: str = Field(default="", description="Departure city or IATA code, e.g. 'London' or 'LHR'")
    destination: str = Field(default="", description="Destination city, e.g. 'Paris'")
    start_date: str = Field(default="", description="Departure date in YYYY-MM-DD")
    end_date: str = Field(default="", description="Return date in YYYY-MM-DD (optional)")
    budget: float = Field(default=0.0, description="Total trip budget in USD")
    travelers: int = Field(default=1, description="Number of adult travelers")
    preferences: list[str] = Field(default_factory=list, description="Trip preferences, e.g. ['adventure', 'food']")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------
def route_from_form(
    origin: str,
    destination: str,
    departure_date: date,
    return_date: date | None,
    budget: float,
    travelers: int,
    preferences: list[str],
) -> TripRequest:
    """Build a TripRequest directly from sidebar form values."""
    return TripRequest(
        origin=origin,
        origin_name=origin,
        destination=destination,
        destination_name=destination,
        start_date=departure_date.isoformat(),
        end_date=return_date.isoformat() if return_date else "",
        budget=float(budget or 0),
        travelers=int(travelers or 1),
        preferences=list(preferences or []),
        currency="USD",
    )


def route_from_text(text: str, llm: Any) -> TripRequest:
    """Extract a TripRequest from free text via LLM structured output.

    The LLM result is merged with the rule-based parser so any field the
    model missed (or got wrong) is filled from the regex fallback.
    """
    rules = _extract_with_rules(text)
    if llm is not None and llm.available:
        extracted = _extract_with_llm(text, llm)
        if extracted:
            return _merge(rules, extracted)
    return rules


# ---------------------------------------------------------------------------
# LLM extraction (pydantic structured output)
# ---------------------------------------------------------------------------
def _extract_with_llm(text: str, llm: Any) -> TripRequest | None:
    try:
        structured = llm.with_structured_output(_TripRequestExtraction)
        if structured is None:
            return None
        result = structured.invoke(
            [
                {"role": "system", "content": _SYSTEM_PROMPT},
                {"role": "user", "content": text},
            ]
        )
        if isinstance(result, _TripRequestExtraction):
            return TripRequest(**result.model_dump())
        if isinstance(result, dict):
            return TripRequest(**result)
    except Exception as exc:  # noqa: BLE001
        print(f"[router] LLM structured extraction failed: {exc}")
    return None


def _merge(rules: TripRequest, llm: TripRequest) -> TripRequest:
    """LLM wins for fields it filled; rules fill the gaps."""
    merged = rules.model_copy(deep=True)
    if llm.origin:
        merged.origin = llm.origin
        merged.origin_name = llm.origin
    if llm.destination:
        merged.destination = llm.destination
        merged.destination_name = llm.destination
    if llm.start_date:
        merged.start_date = llm.start_date
    if llm.end_date:
        merged.end_date = llm.end_date
    if llm.budget:
        merged.budget = llm.budget
    if llm.travelers > 1:
        merged.travelers = llm.travelers
    if llm.preferences:
        merged.preferences = llm.preferences
    return merged


# ---------------------------------------------------------------------------
# Rule-based fallback
# ---------------------------------------------------------------------------
def _extract_with_rules(text: str) -> TripRequest:
    """Regex-based extraction used when no LLM is available."""
    req = TripRequest()

    # Dates: YYYY-MM-DD or DD/MM/YYYY.
    date_matches = re.findall(
        r"\b(\d{4}-\d{2}-\d{2}|\d{1,2}[/-]\d{1,2}[/-]\d{2,4})\b", text
    )
    if date_matches:
        req.start_date = _normalize_date(date_matches[0])
        if len(date_matches) > 1:
            req.end_date = _normalize_date(date_matches[1])

    # Budget: "$1200", "1200 dollars", "budget of 1500".
    budget_match = re.search(
        r"\$\s?(\d+(?:\.\d+)?)|(\d+(?:\.\d+)?)\s*(?:usd|dollars|bucks)",
        text,
        re.IGNORECASE,
    )
    if budget_match:
        req.budget = float(budget_match.group(1) or budget_match.group(2))

    # Travelers: "2 people", "3 adults", "for 4".
    travelers_match = re.search(
        r"(\d+)\s*(?:people|adults|travelers|travellers|persons)", text, re.IGNORECASE
    )
    if travelers_match:
        req.travelers = max(1, int(travelers_match.group(1)))

    # Preferences: known keywords.
    pref_keywords = {
        "adventure": ["adventure", "hiking", "trek", "outdoor"],
        "food": ["food", "culinary", "restaurant", "eat"],
        "culture": ["culture", "museum", "history", "heritage"],
        "beach": ["beach", "sun", "relax"],
        "shopping": ["shopping", "mall"],
        "nightlife": ["nightlife", "party", "club"],
        "nature": ["nature", "park", "scenic"],
        "family": ["family", "kids", "children"],
    }
    for label, keywords in pref_keywords.items():
        if any(k in text.lower() for k in keywords):
            req.preferences.append(label)

    # "from X to Y" pattern for origin/destination.
    from_to = re.search(
        r"\bfrom\s+([A-Za-z\s]{2,30}?)\s+to\s+([A-Za-z\s]{2,30}?)(?:\s|$|,|\.)",
        text,
        re.IGNORECASE,
    )
    if from_to:
        req.origin = from_to.group(1).strip().title()
        req.origin_name = req.origin
        req.destination = from_to.group(2).strip().title()
        req.destination_name = req.destination
    else:
        # Fallback: last capitalized word(s) as destination.
        words = re.findall(r"\b[A-Z][a-z]+\b", text)
        if words:
            req.destination = words[-1]
            req.destination_name = words[-1]

    if not req.start_date:
        req.start_date = (date.today() + timedelta(days=14)).isoformat()
    return req


# ---------------------------------------------------------------------------
# Date helpers
# ---------------------------------------------------------------------------
def _normalize_date(value: Any) -> str:
    """Accept YYYY-MM-DD, DD/MM/YYYY, or relative words; return YYYY-MM-DD."""
    if not value:
        return ""
    value = str(value).strip()
    if re.fullmatch(r"\d{4}-\d{2}-\d{2}", value):
        return value
    match = re.fullmatch(r"(\d{1,2})[/-](\d{1,2})[/-](\d{2,4})", value)
    if match:
        d, m, y = match.groups()
        y = f"20{y}" if len(y) == 2 else y
        try:
            return date(int(y), int(m), int(d)).isoformat()
        except ValueError:
            return ""
    today = date.today()
    rel = {
        "today": today,
        "tomorrow": today + timedelta(days=1),
        "next week": today + timedelta(days=7),
        "next month": today + timedelta(days=30),
    }
    lowered = value.lower()
    for key, d in rel.items():
        if key in lowered:
            return d.isoformat()
    weekday_match = re.search(
        r"next\s+(monday|tuesday|wednesday|thursday|friday|saturday|sunday)", lowered
    )
    if weekday_match:
        target = {
            "monday": 0, "tuesday": 1, "wednesday": 2, "thursday": 3,
            "friday": 4, "saturday": 5, "sunday": 6,
        }[weekday_match.group(1)]
        delta = (target - today.weekday()) % 7
        delta = delta if delta > 0 else 7
        return (today + timedelta(days=delta)).isoformat()
    return ""

File: server/router.ts
import { GoogleGenAI } from '@google/genai';
import { CONFIG } from './config.js';
import { resolveCity } from './geo.js';

export interface TripRequest {
  origin: string;
  origin_name: string;
  destination: string;
  destination_name: string;
  start_date: string;
  end_date: string;
  budget: number;
  travelers: number;
  preferences: string[];
  currency: string;
}

export function routeFromForm(params: {
  origin: string;
  destination: string;
  start_date: string;
  end_date?: string;
  budget?: number;
  travelers?: number;
  preferences?: string[];
}): TripRequest {
  const originInfo = resolveCity(params.origin);
  const destInfo = resolveCity(params.destination);

  return {
    origin: originInfo?.iata || params.origin.trim().toUpperCase(),
    origin_name: originInfo?.name || params.origin.trim(),
    destination: destInfo?.iata || params.destination.trim().toUpperCase(),
    destination_name: destInfo?.name || params.destination.trim(),
    start_date: params.start_date || getDefaultStartDate(),
    end_date: params.end_date || getDefaultEndDate(params.start_date),
    budget: Number(params.budget || 0),
    travelers: Math.max(1, Number(params.travelers || 1)),
    preferences: Array.isArray(params.preferences) ? params.preferences : [],
    currency: 'USD'
  };
}

export async function routeFromText(text: string): Promise<TripRequest> {
  const rules = extractWithRules(text);

  // Attempt LLM structured extraction if GEMINI_API_KEY is available
  if (CONFIG.GEMINI_API_KEY) {
    try {
      const ai = new GoogleGenAI({ apiKey: CONFIG.GEMINI_API_KEY });
      const prompt = `You are an AI Travel Assistant router.
Extract the trip details from this user query: "${text}".
Current date is: ${new Date().toISOString().slice(0, 10)}.
Output valid JSON only matching this schema:
{
  "origin": "Departure city or airport name (e.g. London)",
  "destination": "Destination city name (e.g. Paris)",
  "start_date": "YYYY-MM-DD",
  "end_date": "YYYY-MM-DD or empty string",
  "budget": number in USD,
  "travelers": number of adult travelers,
  "preferences": ["food", "culture", "adventure", "beach", "nightlife", "nature", "shopping", "relaxation"]
}`;

      const response = await ai.models.generateContent({
        model: 'gemini-2.5-flash',
        contents: prompt,
        config: {
          responseMimeType: 'application/json',
          temperature: 0.1,
        }
      });

      if (response.text) {
        const parsed = JSON.parse(response.text);
        return mergeResults(rules, parsed);
      }
    } catch (err) {
      console.warn('[router] Gemini extraction fallback to rules:', err);
    }
  }

  return rules;
}

function mergeResults(rules: TripRequest, llm: any): TripRequest {
  const merged = { ...rules };
  if (llm.origin) {
    const info = resolveCity(llm.origin);
    merged.origin = info?.iata || llm.origin;
    merged.origin_name = info?.name || llm.origin;
  }
  if (llm.destination) {
    const info = resolveCity(llm.destination);
    merged.destination = info?.iata || llm.destination;
    merged.destination_name = info?.name || llm.destination;
  }
  if (llm.start_date && /^\d{4}-\d{2}-\d{2}$/.test(llm.start_date)) {
    merged.start_date = llm.start_date;
  }
  if (llm.end_date && /^\d{4}-\d{2}-\d{2}$/.test(llm.end_date)) {
    merged.end_date = llm.end_date;
  }
  if (llm.budget && Number(llm.budget) > 0) {
    merged.budget = Number(llm.budget);
  }
  if (llm.travelers && Number(llm.travelers) >= 1) {
    merged.travelers = Number(llm.travelers);
  }
  if (Array.isArray(llm.preferences) && llm.preferences.length > 0) {
    merged.preferences = Array.from(new Set([...merged.preferences, ...llm.preferences]));
  }
  return merged;
}

function extractWithRules(text: string): TripRequest {
  let origin = 'London';
  let destination = 'Paris';
  let startDate = '';
  let endDate = '';
  let budget = 0;
  let travelers = 1;
  const preferences: string[] = [];

  // "from X to Y"
  const fromToMatch = text.match(/\bfrom\s+([A-Za-z\s]{2,25}?)\s+to\s+([A-Za-z\s]{2,25}?)(?:\s|$|,|\.)/i);
  if (fromToMatch) {
    origin = fromToMatch[1].trim();
    destination = fromToMatch[2].trim();
  } else {
    // "trip to Paris"
    const toMatch = text.match(/\b(?:to|in|visit|explore)\s+([A-Za-z\s]{2,20}?)(?:\s+for|\s+next|\s+with|\s+in|\s*[,.!?]|$)/i);
    if (toMatch) {
      destination = toMatch[1].trim();
    }
  }

  // Budget matching: "$1500", "1500 dollars", "budget of 2000"
  const budgetMatch = text.match(/\$\s?(\d+(?:\.\d+)?)|(\d+(?:\.\d+)?)\s*(?:usd|dollars|bucks)/i);
  if (budgetMatch) {
    budget = parseFloat(budgetMatch[1] || budgetMatch[2]);
  } else {
    const budgetKeyword = text.match(/budget\s*(?:of|is|:)?\s*\$?(\d+)/i);
    if (budgetKeyword) {
      budget = parseFloat(budgetKeyword[1]);
    }
  }

  // Travelers matching: "2 people", "3 adults", "for 4", "solo", "couple"
  const travelersMatch = text.match(/(\d+)\s*(?:people|adults|travelers|travellers|persons)/i);
  if (travelersMatch) {
    travelers = Math.max(1, parseInt(travelersMatch[1], 10));
  } else if (/\bsolo\b/i.test(text)) {
    travelers = 1;
  } else if (/\bcouple\b/i.test(text)) {
    travelers = 2;
  } else if (/\bfamily\b/i.test(text)) {
    travelers = 4;
  }

  // Preferences keywords
  const prefKeywords: Record<string, string[]> = {
    adventure: ['adventure', 'hiking', 'trek', 'outdoor', 'sports', 'climb'],
    food: ['food', 'culinary', 'restaurant', 'eat', 'dining', 'gastronomy', 'wine', 'pastry'],
    culture: ['culture', 'museum', 'history', 'heritage', 'art', 'historic', 'architecture'],
    beach: ['beach', 'sun', 'swimming', 'ocean', 'coastal'],
    shopping: ['shopping', 'mall', 'boutique', 'fashion', 'market'],
    nightlife: ['nightlife', 'party', 'club', 'bars', 'cocktails'],
    nature: ['nature', 'park', 'scenic', 'garden', 'mountain', 'lake'],
    relaxation: ['relax', 'spa', 'quiet', 'wellness', 'peaceful'],
  };

  const lower = text.toLowerCase();
  for (const [key, words] of Object.entries(prefKeywords)) {
    if (words.some(w => lower.includes(w))) {
      preferences.push(key);
    }
  }

  // Dates matching
  const dateMatches = text.match(/\b(\d{4}-\d{2}-\d{2}|\d{1,2}[/-]\d{1,2}[/-]\d{2,4})\b/g);
  if (dateMatches && dateMatches.length > 0) {
    startDate = normalizeDate(dateMatches[0]);
    if (dateMatches.length > 1) {
      endDate = normalizeDate(dateMatches[1]);
    }
  }

  // Duration in days: "4 days in Paris", "a week", "3-day trip"
  const daysMatch = text.match(/(\d+)\s*[- ]?day/i);
  let tripDays = daysMatch ? parseInt(daysMatch[1], 10) : 4;
  if (/week/i.test(text)) tripDays = 7;

  if (!startDate) {
    startDate = getDefaultStartDate();
  }
  if (!endDate) {
    const s = new Date(startDate);
    s.setDate(s.getDate() + Math.max(1, tripDays));
    endDate = s.toISOString().slice(0, 10);
  }

  const originInfo = resolveCity(origin);
  const destInfo = resolveCity(destination);

  return {
    origin: originInfo?.iata || origin.toUpperCase(),
    origin_name: originInfo?.name || origin,
    destination: destInfo?.iata || destination.toUpperCase(),
    destination_name: destInfo?.name || destination,
    start_date: startDate,
    end_date: endDate,
    budget: budget || (tripDays * 250 * travelers),
    travelers,
    preferences: preferences.length > 0 ? preferences : ['food', 'culture'],
    currency: 'USD'
  };
}

function normalizeDate(val: string): string {
  if (/^\d{4}-\d{2}-\d{2}$/.test(val)) return val;
  const parts = val.split(/[/-]/);
  if (parts.length === 3) {
    let [d, m, y] = parts;
    if (y.length === 2) y = '20' + y;
    return `${y}-${m.padStart(2, '0')}-${d.padStart(2, '0')}`;
  }
  return getDefaultStartDate();
}

function getDefaultStartDate(): string {
  const d = new Date();
  d.setDate(d.getDate() + 14);
  return d.toISOString().slice(0, 10);
}

function getDefaultEndDate(startDate?: string): string {
  const s = startDate ? new Date(startDate) : new Date();
  if (!startDate) s.setDate(s.getDate() + 14);
  s.setDate(s.getDate() + 4);
  return s.toISOString().slice(0, 10);
}
